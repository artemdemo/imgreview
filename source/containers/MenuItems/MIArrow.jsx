import React from 'react';
import { connect } from 'react-redux';
import Icon from '../../components/Icon/Icon';
import MainMenuItem from '../../components/MainMenu/MainMenuItem';
import Arrow from '../../canvas/Arrow/Arrow';
import { addArrow, blurShapes } from '../../model/shapes/shapesActions';

class MIArrow extends React.PureComponent {
    onClick = () => {
        const { shapes, canvas, addArrow, blurShapes } = this.props;
        const arrow = new Arrow({
            stroke: shapes.stroke,
            strokeWidth: shapes.strokeWidth,
        });
        arrow.addToStage(canvas.stage);
        arrow.on('click', arrowInstance => blurShapes(arrowInstance));
        addArrow(arrow);
    };

    render() {
        const { canvas } = this.props;
        return (
            <MainMenuItem
                onClick={this.onClick}
                disabled={canvas.image == null}
            >
                <Icon name='mouse-pointer' />
            </MainMenuItem>
        );
    }
}

export default connect(
    state => ({
        canvas: state.canvas,
        shapes: state.shapes,
    }),
    {
        addArrow,
        blurShapes,
    }
)(MIArrow);
