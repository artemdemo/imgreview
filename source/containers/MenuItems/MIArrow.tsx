import React from 'react';
import Icon from '../../components/Icon/Icon';
import TopMenuItem from '../../components/TopMenu/TopMenuItem';
import { connectArrow } from '../../model/connectShape';

type Props = {
    disabled: boolean;
};

class MIArrow extends React.PureComponent<Props> {
    static readonly defaultProps = {
        disabled: false,
    };

    onClick = () => {
        connectArrow();
    };

    render() {
        const { disabled } = this.props;
        return (
            <TopMenuItem
                onClick={this.onClick}
                disabled={disabled}
            >
                <Icon
                    name='mouse-pointer'
                    title='Arrow'
                />
            </TopMenuItem>
        );
    }
}

export default MIArrow;
