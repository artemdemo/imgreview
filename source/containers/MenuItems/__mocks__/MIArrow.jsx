import React from 'react';

const MIArrow = props => (
    <div data-mock='MIArrow'>
        {props.children}
    </div>
);

export default MIArrow;
