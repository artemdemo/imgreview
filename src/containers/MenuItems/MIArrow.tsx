import React from 'react';
import {connect} from 'react-redux';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faLongArrowAltUp} from '@fortawesome/free-solid-svg-icons';
import TopMenuItem from '../../components/TopMenu/TopMenuItem';
import * as canvasApi from '../../../srcCanvas/api';
import {TReduxState} from '../../reducers';
import {TStateMenu} from '../../model/menu/menuReducer';
import {setShapeToAdd, TSetShapeToAdd} from '../../model/menu/menuActions';
import * as gaService from '../../services/ganalytics';
import { t } from '../../services/i18n';

type TProps = {
    disabled: boolean;
    menu: TStateMenu;
    setShapeToAdd: TSetShapeToAdd;
};

type TState = {
    active: boolean;
};

class MIArrow extends React.PureComponent<TProps, TState> {
    static readonly defaultProps = {
        disabled: false,
    };

    #unsubShapeAdded;

    state = {
        active: false,
    };

    componentDidMount() {
        // @ts-ignore
        this.#unsubShapeAdded = canvasApi.shapeAdded.on(this.handleShapeAdded);
    }

    componentWillUnmount() {
        this.#unsubShapeAdded();
    }

    handleShapeAdded = () => {
        this.setState({
            active: false,
        });
    };

    onClick = () => {
        const { menu, setShapeToAdd } = this.props;
        canvasApi.startAddingShape({
            type: canvasApi.EShapeTypes.ARROW,
            options: {
                strokeColor: menu.strokeColor,
                strokeWidth: menu.strokeWidth,
            },
        });

        setShapeToAdd(canvasApi.EShapeTypes.ARROW);

        gaService.sendEvent({
            eventCategory: gaService.EEventCategories.MenuClick,
            eventAction: gaService.EEventActions.AddArrow,
        });
    };

    render() {
        const { disabled, menu } = this.props;
        return (
            <TopMenuItem
                onClick={this.onClick}
                disabled={disabled}
                active={menu.selectedShapeToAdd === canvasApi.EShapeTypes.ARROW}
                title={t('menu.addArrow')}
            >
                <FontAwesomeIcon icon={faLongArrowAltUp} rotation={270} />
            </TopMenuItem>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        menu: state.menu,
    }), {
        setShapeToAdd,
    },
)(MIArrow);
