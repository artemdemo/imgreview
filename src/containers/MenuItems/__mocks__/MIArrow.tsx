import React from 'react';

const MIArrow = props => (
    <div data-mock='MIArrow'>
        {JSON.stringify(props, null, 2)}
    </div>
);

export default MIArrow;
