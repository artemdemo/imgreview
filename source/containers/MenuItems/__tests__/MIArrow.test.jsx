import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import MIArrow from '../MIArrow.tsx';

jest.mock('react-redux');
jest.mock('../../../canvas/Arrow/Arrow.ts');
jest.mock('../../../components/Icon/Icon');
jest.mock('../../../components/TopMenu/TopMenuItem');
jest.mock('../../../model/connectShape');

describe('MIArrow', () => {
    const connectShapeMock = require('../../../model/connectShape');

    beforeAll(() => {
        jest.clearAllMocks();
    });

    it('should render', () => {
        const tree = renderer.create(
            <MIArrow />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should be disabled', () => {
        const tree = renderer.create(
            <MIArrow
                disabled
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should handle click', () => {
        const wrapper = mount(
            <MIArrow />
        );
        wrapper.simulate('click');
        expect(connectShapeMock.connectArrow).toBeCalled();
    });
});
