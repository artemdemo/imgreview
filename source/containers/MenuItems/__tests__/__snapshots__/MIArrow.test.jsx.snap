// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MIArrow should be disabled 1`] = `
<div
  data-mock="TopMenuItem"
  onClick={[Function]}
>
  <div
    data-mock="Icon"
  />
</div>
`;

exports[`MIArrow should render 1`] = `
<div
  data-mock="TopMenuItem"
  onClick={[Function]}
>
  <div
    data-mock="Icon"
  />
</div>
`;
